pytorch_toolkit/formula_recognition/tools/evaluation_tools.py

```diff
@@ -0,0 +1,385 @@
+import numpy as np
+
+MAX_PX_ROW_DIFF = 3
+MAX_PX_COL_DIFF = 0
+TIMEOUT = 10
+PRINT_FREQ = 100
+
+template = r"""
+\documentclass[12pt]{article}
+\pagestyle{empty}
+\usepackage{amsmath}
+\begin{document}
+\begin{displaymath}
+%s
+\end{displaymath}
+\end{document}
+"""
+
+
+def check_environment():
+    command = subprocess.run(["pdflatex", "--version"], stdout=PIPE, stderr=PIPE, check=True)
+    if command.stderr:
+        raise EnvironmentError("pdflatex not installed, please install it")
+    command = subprocess.run(["gs", "--version"], stdout=PIPE, stderr=PIPE, check=True)
+    if command.stderr:
+        raise EnvironmentError("ghostscript not installed, please install it")
+    command = subprocess.run(["convert", "--version"], stdout=PIPE, stderr=PIPE, check=True)
+    if command.stderr:
+        raise EnvironmentError("imagemagick not installed, please install it")
+
+
+def crop_image(img, output_path, default_size=None):
+    old_im = cv.imread(img, cv.IMREAD_GRAYSCALE)
+    img_data = np.copy(old_im)
+    nnz_inds = np.where(img_data != 255)
+    if len(nnz_inds[0]) == 0:
+        if not default_size:
+            cv.imwrite(output_path, old_im)
+            return False
+        assert len(default_size) == 2, default_size
+        x_min, y_min, x_max, y_max = 0, 0, default_size[0], default_size[1]
+        old_im = old_im[y_min: y_max + 1, x_min, x_max + 1]
+        cv.imwrite(output_path, old_im)
+        return False
+    y_min = np.min(nnz_inds[0])
+    y_max = np.max(nnz_inds[0])
+    x_min = np.min(nnz_inds[1])
+    x_max = np.max(nnz_inds[1])
+
+    old_im = old_im[y_min: y_max + 1, x_min: x_max + 1]
+    cv.imwrite(output_path, old_im)
+    return True
+
+
+def run(cmd, timeout_sec):
+    proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.DEVNULL)
+
+    def kill_proc(process):
+        return process.kill()
+    timer = Timer(timeout_sec, kill_proc, [proc])
+    try:
+        timer.start()
+        _, _ = proc.communicate()
+    finally:
+        timer.cancel()
+
+
+def preprocess_formula(l):
+    """Formula preprocessing
+
+    Args:
+        l (str): Input formula
+
+    Returns:
+        str: Preprocessed formula
+    """
+    l = l.strip()
+    l = l.replace(r'\pmatrix', r'\mypmatrix')
+    l = l.replace(r'\matrix', r'\mymatrix')
+    # remove leading comments
+    l = l.strip('%')
+    if len(l) == 0:
+        l = '\\hspace{1cm}'
+    # \hspace {1 . 5 cm} -> \hspace {1.5cm}
+    for space in ["hspace", "vspace"]:
+        match = re.finditer(space + " {(.*?)}", l)
+        if match:
+            new_l = ""
+            last = 0
+            for m in match:
+                new_l = new_l + l[last:m.start(1)] + m.group(1).replace(" ", "")
+                last = m.end(1)
+            new_l = new_l + l[last:]
+            l = new_l
+    return l
+
+
+def render_routine(line):
+    """Function for rendering single formula
+
+    Args:
+        line (tuple): formula idx, formula string, path to store rendered image
```

Comment:
Vladimir, it seems that this description is not correct -- now the tuple is `(formula, file_idx, folder_path)`

Reply:
Fixed, thanks